### PR TITLE
Preventing imeta mod bypass

### PR DESCRIPTION
**Pull request by FifthPotato**
Seems to work so far. Need to do a bit more research to figure out if there are any other cases I missed.

Need to write tests, too.

### CODE REVIEW

**Comment by trel** [comment]
this is to address #58?

**Comment by trel** [comment]
good so far - let's get a couple tests in today.

File: src/main.cpp [hunk 1]
```diff
@@ -166,10 +166,25 @@ namespace
 			//     {"type": "user",  "name": "jane#otherZone"}
 			//   ]
 			// }
+
+			// Per the API, any of arg5 to arg8 is eligible to be the new attribute
+			// (depending on which one starts with "n:")
+			// However, only one will be considered (multiples will throw CAT_INVALID_ARGUMENT)
+			// Therefore, find the first one that starts with "n:"
+			const auto candidates = {input->arg5, input->arg6, input->arg7, input->arg8};
```
Comment (korydraughn) [inline src/main.cpp:174]: Can you explain why it's four inputs instead of three?

Is it because the units input for the source metadata is optional?

Comment (FifthPotato) [reply src/main.cpp:174]: Referencing https://github.com/irods/irods/blob/e6df49f58759d6609650b82ce4a907d82ee10881/plugins/database/src/db_plugin.cpp#L9444 here, the arguments get passed in top-to-bottom. arg5 is `_unitsOrArg0` which is eligible to be one of the `n:` 
https://github.com/irods/irods/blob/e6df49f58759d6609650b82ce4a907d82ee10881/plugins/database/src/db_plugin.cpp#L9485
...if its type is considered as such. 
https://github.com/irods/irods/blob/e6df49f58759d6609650b82ce4a907d82ee10881/plugins/database/src/db_plugin.cpp#L9499
This shows that those 4 can potentially be the `n:`, so we check all 4 of them. 
https://github.com/irods/irods/blob/e6df49f58759d6609650b82ce4a907d82ee10881/plugins/database/src/db_plugin.cpp#L9502-L9506
This shows that CAT_INVALID_ARGUMENT is thrown if you try to pass `n:` more than once, so just finding the first one should be good enough.

**Comment by korydraughn** [comment]
All we need now are the tests and we're good to go.

**Comment by FifthPotato** [comment]
Added test, and force-pushed because I forgot to update main on my repo after my last PR.

File: packaging/test_rule_engine_plugin_metadata_guard.py [hunk 2]
```diff
@@ -287,6 +287,28 @@ def test_plugin_cannot_be_bypassed_with_addw__issue_40(self):
...
+            'admin_only': True
+        })
+        self.rods.assert_icommand(['imeta', 'set', '-C', root_coll, self.metadata_guard_attribute_name(), json_config])
+
+        try:
+            with lib.file_backed_up(config.server_config_path):
+                self.enable_rule_engine_plugin(config)
+
+                self.user.assert_icommand(['imeta', 'add', '-C', self.user.session_collection, 'unprotected::issue58', 'v'])
+                self.user.assert_icommand(['imeta', 'mod', '-C', self.user.session_collection, 'unprotected::tomato', 'v', 'n:irods::tomato'], 'STDERR', ['CAT_INSUFFICIENT_PRIVILEGE_LEVEL'])
```
Comment (trel) [inline packaging/test_rule_engine_plugin_metadata_guard.py:306]: i think we want to make sure it doesn't update the existing one?

"unprotected::tomato" -> "unprotected::issue58" ?

or perhaps more cleanly...

`'unprotected::issue58', 'v', 'n:irods::issue58'`

Comment (korydraughn) [reply packaging/test_rule_engine_plugin_metadata_guard.py]: @FifthPotato What did you determine? Was the original code a copy-paste issue?

File: packaging/test_rule_engine_plugin_metadata_guard.py [hunk 3]
```diff
@@ -287,6 +287,28 @@ def test_plugin_cannot_be_bypassed_with_addw__issue_40(self):
...
+            'admin_only': True
+        })
+        self.rods.assert_icommand(['imeta', 'set', '-C', root_coll, self.metadata_guard_attribute_name(), json_config])
+
+        try:
+            with lib.file_backed_up(config.server_config_path):
+                self.enable_rule_engine_plugin(config)
+
+                self.user.assert_icommand(['imeta', 'add', '-C', self.user.session_collection, 'unprotected::issue58', 'v'])
+                self.user.assert_icommand(['imeta', 'mod', '-C', self.user.session_collection, 'unprotected::issue58', 'v', 'n:irods::issue58'], 'STDERR', ['CAT_INSUFFICIENT_PRIVILEGE_LEVEL'])
```
Comment (korydraughn) [inline packaging/test_rule_engine_plugin_metadata_guard.py:306]: Let's add another set of assertions that include the units value.

**Review by korydraughn — COMMENTED** [review]
Once the final review comments are resolved. Please run the tests and report whether they succeeded/failed.

File: packaging/test_rule_engine_plugin_metadata_guard.py (hunk 2 above)
Comment (FifthPotato) [reply packaging/test_rule_engine_plugin_metadata_guard.py]: Yep, made a last-minute change that I didn't propagate correctly. Just changed it to what Terrell recommended.

File: packaging/test_rule_engine_plugin_metadata_guard.py [hunk 4]
```diff
@@ -287,6 +287,31 @@ def test_plugin_cannot_be_bypassed_with_addw__issue_40(self):
...
+
+                # Same as above, but with units
+                self.user.assert_icommand(['imeta', 'add', '-C', self.user.session_collection, 'unprotected::withunits', 'v', 'u'])
+                self.user.assert_icommand(['imeta', 'mod', '-C', self.user.session_collection, 'unprotected::withunits', 'v', 'n:irods::withunits', 'u'], 'STDERR', ['CAT_INSUFFICIENT_PRIVILEGE_LEVEL'])
```
Comment (korydraughn) [inline packaging/test_rule_engine_plugin_metadata_guard.py:310]: Let's duplicate L310 and swap the `n:<new_attr>` and `u`. That'll help increase confidence in the implementation of the metadata API.

Let's make the current assertion the last one in the set of duplicates.

**Comment by korydraughn** [comment]
Don't forget to mark comments resolved as you complete them.

File: packaging/test_rule_engine_plugin_metadata_guard.py (hunk 4 above)
Comment (FifthPotato) [reply packaging/test_rule_engine_plugin_metadata_guard.py]: Realized the old L310 wasn't doing what I thought it was doing, so I made it do a units update instead (which is what I want, I think?)
Updated now

File: packaging/test_rule_engine_plugin_metadata_guard.py [hunk 5]
```diff
@@ -307,7 +307,8 @@ def test_plugin_cannot_be_bypassed_via_imeta_mod__issue_58(self):
 
                 # Same as above, but with units
                 self.user.assert_icommand(['imeta', 'add', '-C', self.user.session_collection, 'unprotected::withunits', 'v', 'u'])
-                self.user.assert_icommand(['imeta', 'mod', '-C', self.user.session_collection, 'unprotected::withunits', 'v', 'n:irods::withunits', 'u'], 'STDERR', ['CAT_INSUFFICIENT_PRIVILEGE_LEVEL'])
+                self.user.assert_icommand(['imeta', 'mod', '-C', self.user.session_collection, 'unprotected::withunits', 'v', 'u', 'n:irods::withunits'], 'STDERR', ['CAT_INSUFFICIENT_PRIVILEGE_LEVEL'])
+                self.user.assert_icommand(['imeta', 'mod', '-C', self.user.session_collection, 'unprotected::withunits', 'v', 'n:irods::withunits', 'u:newunits'], 'STDERR', ['CAT_INSUFFICIENT_PRIVILEGE_LEVEL'])
```
Comment (korydraughn) [inline packaging/test_rule_engine_plugin_metadata_guard.py:311]: `u:newunits` -> `u`

We want to capture the fact that the argument positions don't matter to `imeta`.

File: packaging/test_rule_engine_plugin_metadata_guard.py (hunk 4 above)
Comment (korydraughn) [reply packaging/test_rule_engine_plugin_metadata_guard.py]: Hmm. Are we not confident in the behavior/test?

File: packaging/test_rule_engine_plugin_metadata_guard.py (hunk 5 above)
Comment (FifthPotato) [reply packaging/test_rule_engine_plugin_metadata_guard.py:311]: Oh, I see, so it does match the input "units" when placed there. Gotcha, will change!

Comment (korydraughn) [reply packaging/test_rule_engine_plugin_metadata_guard.py]: @FifthPotato Just read your comment here: https://github.com/irods/irods_rule_engine_plugin_metadata_guard/pull/69#discussion_r1495052305

Is my comment about changing `u:newunits` to `u` still relevant?

File: packaging/test_rule_engine_plugin_metadata_guard.py (hunk 4 above)
Comment (FifthPotato) [reply packaging/test_rule_engine_plugin_metadata_guard.py:311]: Nevermind, I mixed up the meaning. Should be resolved now.

File: packaging/test_rule_engine_plugin_metadata_guard.py (hunk 5 above)
Comment (FifthPotato) [reply packaging/test_rule_engine_plugin_metadata_guard.py]: Yep! Bad interpretation on my part, so we want to check the input units can be passed out-of-order, not the output units. Should be good now!

**Review by korydraughn — COMMENTED** [review]
This may be the one.

Please run the tests and let us know if they pass or fail.

**Comment by FifthPotato** [comment]
Tests pass.

**Comment by korydraughn** [comment]
Good. Please squash to taste, no #'s

**Comment by FifthPotato** [comment]
Squashed

**Comment by korydraughn** [comment]
Let's adjust the commit message so that it isn't specific to `imeta`. How about ...

```
[58] Disallow modification of existing protected metadata.
```

Looking at the test again, it appears we're missing tests which prove `imeta mod <protected_avu> <protected_avu>` is covered.

I'll write up an issue for that.

**Comment by FifthPotato** [comment]
Amended!

**Comment by FifthPotato** [comment]
Octothorpe'd

**Comment by korydraughn** [comment]
Please cherry-pick to the 4-3-stable branch, no #'s.

**Comment by korydraughn** [comment]
@FifthPotato ping!

**Pull request merged**